### ISSUE TITLE
getopts complier error

### ISSUE THREAD

**Issue by b1tr4t** [kenpratt/rusty_scheme#5]
this is an error in getopts library, not rusty_scheme, but it breaks this build so...

After:
```
git clone https://github.com/kenpratt/rusty_scheme.git
cd rusty_scheme
cargo run
```
This is the error I get ...

```
$ cargo run
   Compiling getopts v0.2.11
error[E0503]: cannot use `state` because it was mutably borrowed
   --> /home/sean/.cargo/registry/src/github.com-1ecc6299db9ec823/getopts-0.2.11/src/lib.rs:974:33
    |
921 |     let mut machine = |cont: &mut bool, (i, c): (usize, char)| {
    |                       ---------------------------------------- borrow of `state` occurs here
...
925 |         state = match (state, whitespace, limit) {
    |         ----- borrow occurs due to use of `state` in closure
...
974 |     while cont && match state { B | C => true, A => false } {
    |                                 ^ use of borrowed `state`
975 |         machine(&mut cont, (fake_i, ' '));
    |         ------- borrow later used here

For more information about this error, try `rustc --explain E0503`.
error: could not compile `getopts` due to previous error

```

**Comment by b1tr4t** [comment]
I've resolved the issue with:

```
$ cargo update
    Updating crates.io index
    Updating getopts v0.2.11 -> v0.2.21
    Removing libc v0.1.6
    Removing libc v0.2.43
      Adding libc v0.2.102
    Removing log v0.3.1
      Adding unicode-width v0.1.9
```

Sorry, still learning..

**Comment by kenpratt** [comment]
Oh great, glad to hear you got this working. Sorry, this is a pretty old project, so the dependencies may be out of date and I'm not ever all that sure it would run on newer versions of Rust.

**Comment by b1tr4t** [comment]
It runs, but I haven't tried anything more than that.  I did notice it hadn't seen much action for a while, lol..

I happened to read this article.. https://rust-unofficial.github.io/too-many-lists/ and, having written a couple of 'Scheme inspired' domain specific languages myself (one in C and one template heavy one with C++/Boost), I decided that writing another one would be the perfect way to learn Rust!  :D   

I mentioned my plan to a friend, who sent me a book on Rust and a link to your github.  I haven't investigated closely, but I did see that you're using vectors instead of lists.  I'm thinking of using skip lists myself, and sticking to the pointer paradigm, except my pointers will be indexes.  I used mark/sweep garbage collection in my schemes, and that could work nicely with memory regions.

The first 'scheme' I wrote was linked to an Informix ISAM library and I used it for query building and returning data in tagged string representation (much like xml or json), as a component in a prehistoric (real mode) SaaS application.  The second one was a live video compositing system using OpenGL and v4l.  It had actors and message queues as first class types, a bit like erlang.  Both of these were used in production in more or less mission critical applications, so strayed a long way from the academic vision of Scheme as an implementation of the lambda calculus.  I did get continuations and tail calls working tho.

I don't think I have the source code any more, which is annoying.

I also recently worked on a port of the micropython interpreter to some proprietary hardware.  I see there's is a Python in Rust out there somewhere, so maybe it isn't such a dumb idea to use Rust for interpreters, despite the heavy pointer manipulation that most rely on...  I really like the idea of using the guarantees Rust provides in an interpreter, even if it is a struggle to write.

Anyway, I'll have a look around your repo and I'll let you know if I come up with any useful changes.

Cheers,
  Jono
  ***@***.***
